mobile/src/main/java/org/openhab/habdroid/util/HttpClient.java

```diff
@@ -168,6 +132,129 @@ protected Call prepareCall(String url, String method, Map<String, String> additi
+                : new KeyManager[0];
+
+        try {
+            final SSLContext sslContext = SSLContext.getInstance("TLS");
+            sslContext.init(keyManagers, new TrustManager[] { trustManager }, new SecureRandom());
+            final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
+            builder.sslSocketFactory(sslSocketFactory, trustManager);
+        } catch (Exception e) {
+            Log.d(TAG, "Applying certificate trust settings failed", e);
+            builder.sslSocketFactory(mDefaultSocketFactory, trustManager);
+        }
+
+        mClient = builder.build();
+    }
+
+    private static X509TrustManager getDefaultTrustManager() {
+        try {
+            TrustManagerFactory trustManagerFactory =
+                    TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            trustManagerFactory.init((KeyStore) null);
+            TrustManager[] trustManagers = trustManagerFactory.getTrustManagers();
+
+            for (TrustManager trustManager : trustManagers) {
+                if (trustManager instanceof X509TrustManager) {
+                    return (X509TrustManager) trustManager;
+                }
+            }
+        } catch (Exception e) {
+            Log.d(TAG, "Getting default trust manager failed", e);
+        }
+
+        return null;
+    }
+
+    private static class DummyTrustManager implements X509TrustManager {
+        @Override
+        public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        }
+
+        @Override
+        public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        }
+
+        @Override
+        public X509Certificate[] getAcceptedIssuers() {
+            return new X509Certificate[0];
+        }
+    }
+
+    private static class ClientKeyManager implements X509KeyManager {
+        private final static String TAG = ClientKeyManager.class.getSimpleName();
+
+        private Context mContext;
+        private String mAlias;
+
+        public ClientKeyManager(Context context, String alias) {
+            mContext = context.getApplicationContext();
+            mAlias = alias;
+        }
+
+        @Override
+        public String chooseClientAlias(String[] keyType, Principal[] issuers, Socket socket) {
+            Log.d(TAG, "chooseClientAlias - alias: " + mAlias);
+            return mAlias;
+        }
+
+        @Override
+        public String chooseServerAlias(String keyType, Principal[] issuers, Socket socket) {
+            Log.d(TAG, "chooseServerAlias");
+            return null;
+        }
+
+        @Override
+        public X509Certificate[] getCertificateChain(String alias) {
+            Log.d(TAG, "getCertificateChain", new Throwable());
+            try {
+                return KeyChain.getCertificateChain(mContext, alias);
```

Comment:
Yes. It will be called during socket communication from one of the OkHttp threads. Specifically, it's invoked from an OpenSSL callback which is fired when OpenSSL reads a client certificate request from the socket (I checked this in the sources: https://android.googlesource.com/platform/external/conscrypt/)

I wasn't able to actually check this though (callback wasn't called), probably due to my server configuration. Can you tell me what I'd need to configure on the server to enable client cert authentication - or alternatively - can you give this patch a try? The `new Throwable()` calls I accidentally left in should show the call stacktrace in logcat whenever it's called and should allow to clearly see from where it's called.

Reply:
Unfortunately there is no tutorial how to use client certificate, you can see at https://github.com/openhab/openhab-docs/issues/656


When you finish your work, let me know I can test it.